From f4527a3da0b3670f0249e223b4954a39c4f95fba Mon Sep 17 00:00:00 2001
From: Andreas Oberritter <obi@opendreambox.org>
Date: Fri, 15 Aug 2014 16:39:02 +0200
Subject: [PATCH 9/9] dd: behave like GNU coreutils with conv=fsync

Call it once after everything has been written, otherwise
it's a very big performance hit.

Signed-off-by: Andreas Oberritter <obi@opendreambox.org>
---
 coreutils/dd.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/coreutils/dd.c b/coreutils/dd.c
index 2838f63..3366e2d 100644
--- a/coreutils/dd.c
+++ b/coreutils/dd.c
@@ -445,17 +445,19 @@ int dd_main(int argc UNUSED_PARAM, char **argv)
 			if (write_and_stats(ibuf, n, obs, outfile))
 				goto out_status;
 		}
-
-		if (flags & FLAG_FSYNC) {
-			if (fsync(ofd) < 0)
-				goto die_outfile;
-		}
 	}
 
 	if (ENABLE_FEATURE_DD_IBS_OBS && oc) {
 		if (write_and_stats(obuf, oc, obs, outfile))
 			goto out_status;
 	}
+
+	if (flags & FLAG_FSYNC) {
+		while (fsync(ofd) != 0)
+			if (errno != EINTR)
+				goto die_outfile;
+	}
+
 	if (close(ifd) < 0) {
  die_infile:
 		bb_simple_perror_msg_and_die(infile);
-- 
1.9.1

